### DISCUSSION TITLE
Feature Wishlist

### DISCUSSION THREAD

**Discussion by nebhead** [nebhead/PiFire#27] [Ideas]
Post your feature requests here.  Things that I am planning on working on in the future:

- Support for changing from Fahrenheit to Celsius
- Evaluate using threading for control.py
- Evaluate a blocking library for file access (prevent collisions between app.py and control.py)

**Comment by nebhead** [comment]
Oh yes, and I would also like to fix the settings.json system so that when an update occurs, it overlays the updated settings on the older settings file.  This is more of a bug fix, but also kind of a new feature.

**Comment by weberbox** [comment]
Native Android app
4th probe support
Maybe touchscreen? 

Would be great if settings changes took affect during the current cook. Maybe the control.py threading can help with that?

**Reply by nebhead** [reply]
> Native Android app

Yes!!  I need to fix the API support (it's pretty badly broken at the moment).  Also, I need to learn Android App programming. :)  Definitely on the list.  If anyone lurking here is a pro at Android app development, guidance would be awesome.  

> 4th probe support

I'll add it to the list! 

> Maybe touchscreen?

Yep.  Would be great. 

> Would be great if settings changes took affect during the current cook. Maybe the control.py threading can help with that?

Some settings will take during the cook, but not all of them - i.e. Hold Cycle PID settings might require some changes.  So it's possible (even without the threading), but there would need to be some changes in control.py for sure.

**Reply by weberbox** [reply]
> Yes!! I need to fix the API support (it's pretty badly broken at the moment). Also, I need to learn Android App programming. :) Definitely on the list. If anyone lurking here is a pro at Android app development, guidance would be awesome.

I have some experience but it has been a while and I only know the java side not the newer Kotlin. I have been thinking about trying something here if I get some extra time. If I ever get started I will let you know. 


> Some settings will take during the cook, but not all of them - i.e. Hold Cycle PID settings might require some changes. So it's possible (even without the threading), but there would need to be some changes in control.py for sure.

I noticed that P-Mode does not seem to change along with Smoke Plus settings like max temp etc. needed to switch to a different mode then back. 

I added a semi janky [second grill probe option](https://github.com/zippy-man/PiFire/tree/grill-probe2) since I have the fourth probe added to the ADC and wanted to be able to quickly switch to a thermoworks gill level probe without having to ssh in but that also does not seem to change while in a cook unless it is restarted. It appears that app.py cannot write to control.json when control is active during a cook and control.py does not want to read new settings in settings.json unless the cycle is refreshed. I tried a couple things to correct the on the fly probe change but have not been successful yet.

**Reply by weberbox** [reply]
Ok I have been working on a native Android app for a little while and I think it is at the point where I can share it. I have tested decently on the prototype server and have loaded the new server version onto my hardware. I have run the grill a couple times without issue but more testing should be done on hardware just to be sure. 

First note - I am not an Android developer or software engineer. I only created this in my free time and it probably has issues somewhere. The code is probably spaghetti to a real dev but it is my skill level. Hopefully by sharing it will be useful and more experienced people can help improve or correct mistakes.  It is a good starting point

The server needed a huge update to support communication and I decided to use SocketIO with websockets to support near real time updates without the need to poll the server constantly for dash data. 

Here are some of the changes I made:
Added SocketIO for Android mobile app support
Updated notification settings and added Firebase push notification support
Added option to enable and disable notification types
Added qrcode support to admin page and ili9341 display
Changed from RPi GPIO to gpiozero for ili buttons to support long press --> long press down to show a QR code on the display
gpiozero also improved button responsiveness
Added PelletDB log clear option to Admin Page
Added settings and pelletdb update timestamp for future options
Moved Pellet Level sensor settings to settings.json from pelletdb.json --> I felt all settings should be in settings.json this also allows you to clear the pelletdb and your sensor settings are not wiped. 
Merge in Redis support updates
Merge distance sensor update
Bugfix Smoke Plus toggle in settings
Reformat settings html
Changed history time format to HH:mm:ss

I have two versions of the server. One is based on the normal 3 probes and one for the pcb I created with 4 probes. In the 4 probe version you can just select the stock grill probe or an externally connected one. 

Normal version [here](https://github.com/weberbox/PiFire/tree/mobile-dev-redis-normal)

Multi Grill Probe version [here](https://github.com/weberbox/PiFire/tree/mobile-dev-redis-grillprobes)

You can find the repo for the Android app [here](https://github.com/weberbox/PiFire-Android)

Here are some screenshots 

![dashboard](https://user-images.githubusercontent.com/313328/134221374-25bb424d-27f8-4358-88f1-207af3a573a9.png)
![pelletmanager](https://user-images.githubusercontent.com/313328/134221414-49fb33fc-b38a-4b1c-ac1e-5482fbd37d0c.png)
![eventlog](https://user-images.githubusercontent.com/313328/134221445-6b8cc163-09b5-4e49-a68a-88fdbf570b56.png)
![history](https://user-images.githubusercontent.com/313328/134221467-2b04f3c5-e008-4110-a7e9-506d3479e1bb.png)
![settings](https://user-images.githubusercontent.com/313328/134221479-35f571e2-7c15-4d12-8de5-0ae172aeac3f.png)
![admin](https://user-images.githubusercontent.com/313328/134221499-2c14e842-f6fb-41a6-bad8-3f1a16a23b9a.png)
![workmode](https://user-images.githubusercontent.com/313328/134221519-0d4724af-c0fb-4927-9cb3-1773dc8da63e.png)
![drawer](https://user-images.githubusercontent.com/313328/134221531-7893028b-9bb7-4020-9824-c3b5c2936044.png)
![info](https://user-images.githubusercontent.com/313328/134221541-cf67be87-06d9-438b-acd6-542fade139a5.png)
![tempicker](https://user-images.githubusercontent.com/313328/134221557-6249f604-4065-4e3d-939b-4388cb855cff.png)

**Reply by nebhead** [reply]
@weberbox just WOW!!!  

I am so very impressed by this!  Outstanding work.  I've been looking over the changes on your fork and they are looking great.  I'd say, maybe let's just merge these changes into the development branch of PiFire at the very least.  I wouldn't love to get my hands on the Android app and start testing it out.  

Really very cool - and I'm sure several other folks on here would be interested in trying out the app too.

**Reply by nebhead** [reply]
BTW - I wanted to ask, if you have run into an issues with the threading library and Gunicorn?  A long time ago, I attempted to implement a flask app, together with threading.  It worked OK until I wanted to run Gunicorn, and the whole thing sort of broke.  I never did go back and figure it out.  However, this was with Python 2.7 and a much older version on Gunicorn.  So, just curious if it's all playing nicely in one process now?  If that's the case... eventually, we could just move the control logic back into the main app process.  That would be super cool because then we wouldn't have to do gymnastics with the settings / control files anymore.

**Reply by weberbox** [reply]
@nebhead Sure you can grab the source and compile the app with Android Studio. You would need to do this if you want to use Firebase push notifications as you would need your own google-services.json file and a Firebase project setup first. I just made a couple pushes with new product flavors so it can be complied easier without needing Firebase and not having to remove all the code and dependencies if you do not plan on using it. 

The app does not do much without a connection to the server just FYI. I would like to someday make it so the settings and pellet manager work offline and sync when back online, I added a timestamp to the server but have not worked on it yet on the Android side.

I have been running threading and Gunicorn on the test server for about a month or so and have not experienced any issues so far. I have limited testing on actual hardware so hopefully it does not become an issue. 

I do not like the way I did the QRCode on the ili display as it is blocking and cannot be exited. I did not spend a lot of time on it so maybe you could look at it and see if there is a better way? 

I will make a request to the dev branch. If you play around with it let me know what you think.

**Reply by nebhead** [reply]
FYI... merged all the changes into the development branch, thank you for the pull request!  

For the Android development side, maybe you can start to post the APK's in the "release" section of GitHub?  This allows you to attach stuff to your repository, without disrupting the size of your overall repo or cluttering it with large binaries.  

I'll take a look at the QR Code implementation.  This was something I really wanted to do previously, so I'm glad you actually took the time to do it.  

Also, I am thinking, since you have a 4-probe setup, that I would try to re-architect things a little with respect to how we store the probe history and display it, such that it's better integrated.  This way, in the future, we can all be working off of the same branch.  Also, I feel I have to give back, considering all the work you have done to get the Android side of things going. :)

**Reply by weberbox** [reply]
I did not really plan on having releases as the firebase side requires the user to compile themselves but I guess I could release the non firebase version. It would be nice if there was a way to have firebase without the need for each person to set this up but it would require an app server that stores everyone's app id and ties it to their local pirfre installation with tokens or an api. This would be nice but would require an external server and quite a bit of work to setup. Setting up Firebase and compiling the app is not that hard so it was not worth the effort. 

As for the 3 vs 4 probes it would be nice not having to deal with multiple branches and versions. Currently I am only displaying one grill probe at a time with a selection for type as they are different (stock vs thermoworks). This was the best option for me so the entire UI on the webpage and display required no changes. I am fine with it remaining this way as it would be way easier to update the main branch to support that. Actually I just made a few updates so the Android app will look at the server for 3 or 4 probes and show the options accordingly. This way there does not need to be multiple versions of the app for 3 or 4 probes. 

Do not feel obligated to work on the four probe version just for me. My app is useless without your awesome project so you have already given but I will not refuse working out a way for the main branch to support both. If you want take a look at my four probe version and see if it poses any issues. It might be as easy as just checking if the server is a four probe version or three probe and then showing the relevant options. 

An example might be like below just check

```
if (settings['globals']['four_probes']):
		if (settings['grill_probe_settings']['grill_probe_enabled'][0] == 1):
			AvgGT = adc_data['Grill1Temp']
		else:
			AvgGT = adc_data['Grill2Temp']
	else:
		AvgGT = adc_data['Grill1Temp']
```

But maybe there is a better way than a bunch of if else

**Comment by SeanDMaloney** [comment]
Hello all! First off, I wanted to say how impressive your project is, and thank you for your efforts. 

I had started working on something similar using @DBorello’s project as a framework a couple years ago, but was diverted by life towards other projects. Your move away from Firebase was something that I had wanted to pursue as well, but I lack the experience that you have in making that happen. 

I am looking at doing something similar with a couple of smokers that I have, and am curious to see what it may take to implement the ability to control multiple smokers simultaneously with the GUI interface you have created, or something similar to it. Building in a way to have profiles for different smokers would be something I would be interested in as well, given that I have several smokers that would require different durations of auger run times in order to maintain temperatures and sustain the fires. Two of my three smokers I am working on have augers that feed, more or less, directly into a burn pot, rather than a feed tube like the Traeger.

**Reply by nebhead** [reply]
Thank you @SeanDMaloney!  

Basically same story here!  I started with the @DBorello PiSmoker project and decided to go all in on revamping it without Firebase.  I actually reached out to him and discussed the project and he was extremely helpful and supportive in getting PiFire going.  

As for controlling multiple smokers at the same time, it's not out of the question.  Performance of the Pi, might be limiting in this respect.  I did have some challenges with performance on the Pi Zero W in the beginning, but for the most part I think it's relatively sorted out.  There is still more to do to improve.  I have some plans to move the control, temperature, and history to a built-in Redis server for faster messaging that avoids collisions - which _might_ really turbo-charge performance.  

The other limitation might be with regard to the amount of IO on the platform.  Things might have to be streamlined to ensure there is enough GPIO to support multiple platforms at the same time.  I mean, it's doable but it would be tough.  

My thought would be, perhaps it's best to just have separate Pi Zero's for each smoker, with different tuning for each.  Then using the API (once I get it properly coded up) create another WebUI that aggregates all of the different controllers into one view.

**Reply by SeanDMaloney** [reply]
Multiple devices had always been my conclusion as well for the reasons you stated, as well as allowing a higher level of independence between smokers. 

I’m also curious to see what would be involved in developing an iOS native app that could potentially utilize your API. Working up something that would be able to monitor and control the smokers outside of my local network is my long term goal, preferably without using a VPN, but that’s a ways out.

**Reply by nebhead** [reply]
OK, cool, I think it would just come down to where you run the WebUI for aggregating the data for all of the smokers at once.  Something to consider as I'm sure that would come in handy for folks at competitions, etc.  

Once I get it fixed up, the API would be perfect for using with an Android, iOS, or PC application.  Unfortunately, I don't have any expertise in creating an iOS application myself and I don't own any Apple devices (aside from some super-old ipods).  I would welcome anyone to take a crack at this.  

As for accessing outside the network without VPN, the current web interface isn't secure by any measure.  There is no user/password management or security otherwise built in.  It's something we could implement, but would be quite a bit of work.  If we were just exposing the API, that could be done relatively easily utilizing the built-in nginx reverse proxy.  However, again, it's a little bit risky given you've got things that make fire, exposed to the internet. :)

**Reply by DBorello** [reply]
Greetings All!

I am glad to hear the smoker control community is alive and well!

I am always delighted to see the derivatives of PiSmoker.  It was a fun little hacky project 6 years ago, with almost no progress since.  However, it is still running strong.

I can't blame you for dropping Firebase.  I am not a fan of the external dependency on a third party.  However, it has been convenient to not have to mess with port forwarding, etc.  I like the comments on exposing fire devices to the internet :)  I do recall using the same Pi (whoops) for development of another project.  I didn't realize I was toggling GPIOs connected to an igniter.  

Also, can someone please write a better controller than my simple PID :)

**Reply by nebhead** [reply]
@DBorello!  Thanks for dropping in on the conversation!  

Your PiSmoker project has definitely been an inspiration for me and for many other folks out there.  I basically just put a different WebUI on it. :)  

And I feel your PID works pretty well as-is.  In fact, I'd say it's better than my original Traeger controller by a mile.  I'd toyed around with the idea of trying a few other PID modules that are floating around on GitHub, but hadn't gotten around to it yet.  I lack experience with PID's in general to really drill down into the design.  With that said I do think there may be some lurkers here in this project that have more experience and could potentially help out to make it better.

**Reply by SeanDMaloney** [reply]
I couldn’t agree more about the potential security risks associated with an unprotected combustion appliance exposed to the masses. I would definitely be implementing some sort of security protocols to mitigate those concerns, as I have no desire to create a spontaneously igniting smoker or grill. My workshop has enough fire hazards (acetylene hoses that decide the flame comes out the hose instead of torch, just to name one) without eliciting outside assistance!

@DBorello I am happy to see you are still active, given your project gave me the confidence and groundwork to undertake this myself. 

While my current modern programming experience is vastly dwarfed by most of the contributors here, I do have a tremendous amount of practical experience rebuilding and troubleshooting most all things pellet smoker. I have all of the currently produced Traeger controllers (Pro, Arc, and AC version of the WiFire controller), Louisiana controllers, and a Pit Boss controller for my vertical cabinet smoker. I also have a fairly developed knowledge of the workings of the new DC controllers that Traeger is using, so if there are any questions, details, or insights that I can offer, it would be my privilege to do so.
